### PR TITLE
Wrapping concurrent objects

### LINKED ISSUE #669: [core] High CPU load on Windows
**OS Host:** Windows 10 64 bits
**SRT version:** latest git (f5f52c02c27d7a8d7050a746eebee6e3d1b33aaa)
**Program:** srt-live-transmit.exe (32 bits compile)

Commandline used:
`srt-live-transmit "srt://wowza_server....." udp://127.0.0.1:1234 -s:10`

It starts using about 0.2% of CPU load. After some hours (8 or even more), it jumps to about 70% of CPU load. However memory usage keeps stable at 2 MB.

How can I capture the moment when this issue happens ?
I have my toolchain ready for anything you want me to compile.

Regards

### PR DESCRIPTION
**Pull request by maxsharabayko**
This PR is the second out of four expected PRs for SRT library to work with C++11 threading library, and still being able to use pthreads (controlled by `USE_STL_CHRONO` macro definition).

The first PR was #929. It introduced steady clock time wrappers, that implement the same API, as chrono clock. Thus it can be mapped directly to STL chrono version if C++11 is selected.
This PR wraps concurrent objects (mutexes and CVs) to be able to use C++11 version.
The next PR will wrap threads and thread local storage (TLS).

## List of changes

* Created a wrapper class for Mutex to wrap pthreads mutex and map C++11 mutex.
* Created `SyncEvent` class to wrap conditional variables.

## Concurrency Control

To be able to use both `pthread_mutex_t` and `std::threads::mutex`, an abstraction is required.

In case of C++11 `std::threads::mutex`, being a class, can be used directly. The only concern is that operations on a C++11 mutex can throw exceptions.

In case of C++03 and `pthreads`, the new `Mutex` class wraps a `pthread_mutex_t` and provides the API similar to the API of the C++11 mutex.

The `SyncEvent` class is introduced to represent a synchronization event. Instances of `SyncEvent` replace conditional variables and `CTimer::sleepto()`.
Also SRT core has a global synchronization timer, accessed via `CTimer::waitForEvent()` (actually waits for 10 ms) and `CTimer::triggerEvent()`. Those functions were used to synchronize state changes across the whole library. This PR instead uses a static instance of `SyncEvent` for these purposes.

* `CTimer::triggerEvent()` -> `SyncEvent::notify_one()`;

* `CTimer::waitForEvent()` -> `SyncEvent::wait_for(10ms)`.

Closes #577 (C++11 mutexes) as replaced.

### TODO

- [ ] Busy waiting on a c++11 SyncEvent. If applied to the whole class, will impact sending pace, as well as reading timing. Probably need to template the `SyncEvent` to allow busy waiting or not.
- [ ] Operations on a C++11 mutex can throw an exception. `Mutex` class should probably catch them?
- [ ] Problems with `PTHREAD_COND_INITIALIZER` of a `SyncEvent` and object destruction. If created with `pthread_cond_init()` destructors of `s_SyncEvent` and `CUDTUnited::m_GCState` are called twice (at least on Windows).
- [x] Move static `s_SyncEvent` inside the `sync.cpp`
- [ ] Bear in mind #732 and #733 (monotonic clock for garbage collector).
- [ ] Rename namespace `srt_logging` to `srt::logging`.

### Update 20.12.2019

The initial version of this PR was moved to develop/sync_01_timers_after_pr929
This PR is updated with the changes from PR #929, which introduced timers.

### Update 20.01.2020

Merged a subset of these changes with PR #1064.

### Update 18.03.2020

This PR was split into the following smaller PRs and can be closed.

- [X] **Mutex wrapper class** (PR #1064) to switch between pthread and C++11 implementations.
- [X] **CV waiting functions** (PR #1064). Waiting on a CV has to be wrapped into some function to allow switching between C++11 and pthread versions.
- [X] To avoid merge conflicts with the network bonding branch, the `CSync` wrapper was added as an intermediate step (PR #1067). For the moment it allows to simplify merging due to this class was already added in that branch. It is to be replaced by `CEvent` and `CTimer` classes.
- [x] **Condition variable wrapper** (PR #1102)
- [x] **`CEvent` class** to hold both mutex and CV. To gradually replace `CSync` in further changes. PR #1149 
- [x] **Implementation of `CTimer`** based on `CEvent`. PR #1174.
- [x] Monotonic clock to be used in all sync events when enabled. PR #1174.

### CODE REVIEW

**Comment by justinjoy** [comment]
Does that mean C++11 should be supported from 1.4 ?

**Comment by maxlovic** [comment]
@justinjoy Hopefully from 1.4.1.

**Comment by justinjoy** [comment]
> @justinjoy Hopefully from 1.4.1.

Before enabling C++11 by default, you or someone else should check if it is possible to remove from VLC contrib.

http://git.videolan.org/?p=vlc.git;a=blob;f=contrib/src/srt/rules.mak;h=c9dbb59db87ecd5c9c87ff32cd8900ae4d759a68;hb=HEAD#l40

As I remember, I should turn off because vlc asked.

File: srtcore/buffer.cpp [hunk 1]
```diff
@@ -187,7 +181,7 @@ void CSndBuffer::addBuffer(const char* data, int len, int ttl, bool order, uint6
         // [PB_SOLO] - 1 packet per message
 
         s->m_ullSourceTime_us = srctime;
```
Comment (ethouris) [inline srtcore/buffer.cpp]: Shouldn't this SourceTime be also of type `time_point`? Note that this is a time INTERNALLY calculated.

File: srtcore/buffer.cpp [hunk 2]
```diff
@@ -196,19 +190,19 @@ void CSndBuffer::addBuffer(const char* data, int len, int ttl, bool order, uint6
     }
     m_pLastBlock = s;
 
-    CGuard::enterCS(m_BufLock);
+    CriticalSection::enter(m_BufLock);
```
Comment (ethouris) [inline srtcore/buffer.cpp]: You have no unusual path here. Maybe better to mark a critical section block?

Comment (maxlovic) [reply srtcore/buffer.cpp]: Do you mean the `ScopedLock` (aka `CGuard`) inside a code block?

File: srtcore/core.cpp [hunk 3]
```diff
@@ -5490,15 +5490,15 @@ int CUDT::receiveBuffer(char *data, int len)
...
             else
             {
                 const steady_clock::time_point exptime = steady_clock::now() + milliseconds_from(m_iRcvTimeOut);
                 while (stillConnected() && !m_pRcvBuffer->isRcvDataReady())
                 {
-                    SyncEvent::wait_for(&m_RecvDataCond, &m_RecvLock, milliseconds_from(m_iRcvTimeOut));
+                    m_RecvDataSync.wait_for(milliseconds_from(m_iRcvTimeOut));
                     if (steady_clock::now() >= exptime)
                         break;
                 }
```
Comment (maxsharabayko) [inline srtcore/core.cpp]: As noted by @ethouris, might be worth replacing with `wait_until(exptime)`.

File: srtcore/queue.cpp [hunk 4]
```diff
@@ -1257,11 +1235,12 @@ static string PacketInfo(const CPacket &pkt)
...
-    // This might be not really necessary, and probably
-    // not good for extensive bidirectional communication.
-    m_pTimer->tick();
-#endif
+    // TODO: Put this change in a separate commit
+    //#if !USE_BUSY_WAITING
+    //    // This might be not really necessary, and probably
+    //    // not good for extensive bidirectional communication.
+    //    m_pTimer->tick();
+    //#endif
```
Comment (maxsharabayko) [inline srtcore/queue.cpp]: - [ ] TODO: Put this change in a separate commit. It changes the synchronization logic.

**Comment by maxsharabayko** [comment]
This PR was split into the following smaller PRs and can be closed.
